### ISSUE TITLE
@babel/react: useSpread needlessly clones props

### ISSUE THREAD

**Issue by nstepien** [babel/babel#11419]
## Bug Report

- [ ] I would like to work on a fix!

**Current Behavior**
Enabling `useSpread` needlessly clones props:
```js
React.createElement(SomeComponent, { ...props })
```
With `useSpread` disabled, you can see that the object is not cloned:
```js
React.createElement(SomeComponent, props)
```

**Input Code**
- REPL or Repo link if applicable:

```jsx
function MyComponent(props) {
  return <SomeComponent {...props} />;
}
```
I can't find a toggle for `useSpread` in the babel REPL. 🤷 

**Expected behavior/code**
When a component only has 1 spread object as prop, no other props/second spread, there is no need to clone the object.

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
- Filename: `babel.config.json`

```js
{
  "presets": [
    ["@babel/env", {
      "bugfixes": true,
      "corejs": 3,
      "useBuiltIns": "entry"
    }],
    "@babel/typescript",
    ["@babel/react", { "useSpread": true }]
  ],
  "plugins": [
    ["@babel/transform-runtime", { "useESModules": true }],
    "@babel/proposal-class-properties"
  ]
}
```

**Environment**

- Babel version(s): [e.g. v6.0.0, v7.0.0-beta.34] `^7.9.4`
- Node/npm version: [e.g. Node 8/npm 5] Node 13.12.0/npm 6.14.4
- OS: [e.g. OSX 10.13.4, Windows 10] Win10
- Monorepo: [e.g. yes/no/Lerna] no
- How you are using Babel: [e.g. `cli`, `register`, `loader`] `loader`

**Possible Solution**

Avoid extra clonings.

**Additional context/Screenshots**
Add any other context about the problem here. If applicable, add screenshots to help explain.
None.

**Comment by nicolo-ribaudo** [comment]
Does cloning props cause any observable effect, maybe in the number of re-renders?

**Comment by nstepien** [comment]
Not that I know of. I just figured we could avoid extra work/bytes when possible.

**Comment by nicolo-ribaudo** [comment]
This affects both the `classic` and `automatic` runtimes. Before going ahead, I would wait for a :+1: from @babel/react

**Comment by nstepien** [comment]
The [`@babel/preset-react` docs](https://babeljs.io/docs/en/babel-preset-react#react-classic-runtime) only list `useSpread` under the `classic` runtime options. 🤔

**Comment by nicolo-ribaudo** [comment]
Yeah, because with `automatic` it's enabled by default :stuck_out_tongue:

Classic output:
```js
function MyComponent(props) {
  return /*#__PURE__*/React.createElement(SomeComponent, { ...props
  });
}
```

Automatic output:
```js
import { jsx as _jsx } from "react/jsx-runtime";

function MyComponent(props) {
  return /*#__PURE__*/_jsx(SomeComponent, { ...props
  });
}
```

**Comment by lunaruan** [comment]
Hey! Cloning props was an intentional decision so we shouldn't change this. @sebmarkbage

**Comment by trueadm** [comment]
Cloning the props is an important heuristic here.

It means that the object that gets passed to React's `jsx` is a plain mutable object with simple properties; otherwise if you passed in `props` directly there are no guarantees.

Some of the properties might not be enumerable, or the properties might be getters/setters, or the object might be frozen/sealed (meaning React would have to clone it again if it wanted to mutate the object). By cloning it, we get a snapshot that is final.

**Comment by jridgewell** [comment]
Closing as intentional per https://github.com/babel/babel/issues/11419#issuecomment-613649672.

**Comment by nstepien** [comment]
@trueadm What about `React.createElement`? Babel doesn't clone the props object in classic mode with `useSpread` disabled, so we could retain that behavior in this case, right?

**Comment by trueadm** [comment]
For ‘createElement’, we don’t need to clone the props as we create another object internally that acts as a clone.

**Comment by nstepien** [comment]
@jridgewell I guess we can reopen and only address the `createElement` case then.

**Comment by gaearon** [comment]
To be clear, we'd like to freeze passed props directly in the future but it would be a breaking change so can't do that now.

**Comment by nicolo-ribaudo** [comment]
Given the comments above, and the fact that even with the classic runtime sometimes Babel compiles it to `clone`, I would:
- suggest you to go ahead with cloning
- update Babel to always clone with the classing runtime (starting from Babel 8?)

**Comment by remcohaszing** [comment]
I would actually take it a step further and always use spread in the Babel transform. People can use `@babel/plugin-transform-spread` if they want to compile to `Object.assign()`.

**Comment by nicolo-ribaudo** [comment]
I agree 👍 We can start doing it in Babel 8.
